packages/user-operation-controller/src/utils/gas-fees.ts

```diff
@@ -0,0 +1,295 @@
+ * @param suggestedGasFees - The suggested gas fees, if any.
+ * @param transaction - The transaction that created the user operation.
+ * @returns The userFeeLevel for the user operation.
+ */
+function getUserFeeLevel(
+  metadata: UserOperationMetadata,
+  originalRequest: AddUserOperationRequest,
+  suggestedGasFees?: SuggestedGasFees,
+  transaction?: TransactionParams,
+): UserFeeLevel {
+  const { origin } = metadata;
+  const { maxFeePerGas, maxPriorityFeePerGas } = originalRequest;
+
+  const {
+    maxFeePerGas: suggestedMaxFeePerGas,
+    maxPriorityFeePerGas: suggestedMaxPriorityFeePerGas,
+  } = suggestedGasFees || {};
+
+  if (
+    isGasFeeEmpty(maxFeePerGas) &&
+    isGasFeeEmpty(maxPriorityFeePerGas) &&
+    transaction?.gasPrice
+  ) {
+    return origin === ORIGIN_METAMASK
+      ? UserFeeLevel.CUSTOM
+      : UserFeeLevel.DAPP_SUGGESTED;
+  }
+
+  if (
+    isGasFeeEmpty(maxFeePerGas) &&
+    isGasFeeEmpty(maxPriorityFeePerGas) &&
+    suggestedMaxFeePerGas &&
+    suggestedMaxPriorityFeePerGas
+  ) {
+    return UserFeeLevel.MEDIUM;
+  }
+
+  if (origin === ORIGIN_METAMASK) {
+    return UserFeeLevel.CUSTOM;
+  }
+
+  return UserFeeLevel.DAPP_SUGGESTED;
+}
+
+/**
+ * Gets suggested gas fees.
+ * @param request - The request to update the gas fees.
+ * @param request.getGasFeeEstimates - A callback to get gas fee estimates.
+ * @param request.provider - A provider to query the network.
+ * @returns The suggested gas fees.
+ */
+async function getSuggestedGasFees(request: UpdateGasFeesRequest) {
+  const { getGasFeeEstimates, provider } = request;
+
+  try {
+    const { gasFeeEstimates, gasEstimateType } = await getGasFeeEstimates();
+
+    if (gasEstimateType === GAS_ESTIMATE_TYPES.FEE_MARKET) {
+      /* istanbul ignore next */
+      const {
+        medium: { suggestedMaxPriorityFeePerGas, suggestedMaxFeePerGas } = {},
+      } = gasFeeEstimates;
+
+      if (suggestedMaxPriorityFeePerGas && suggestedMaxFeePerGas) {
+        const values = {
+          maxFeePerGas: gweiDecimalToWeiHex(suggestedMaxFeePerGas),
+          maxPriorityFeePerGas: gweiDecimalToWeiHex(
+            suggestedMaxPriorityFeePerGas,
+          ),
+        };
+
+        log('Using medium values from fee market estimate', values);
+
+        return values;
+      }
+    }
+
+    if (gasEstimateType === GAS_ESTIMATE_TYPES.LEGACY) {
+      const maxFeePerGas = gweiDecimalToWeiHex(gasFeeEstimates.medium);
+
+      log('Using medium value from legacy estimate', maxFeePerGas);
+
+      return {
+        maxFeePerGas,
+      };
+    }
+
+    if (gasEstimateType === GAS_ESTIMATE_TYPES.ETH_GASPRICE) {
+      const maxFeePerGas = gweiDecimalToWeiHex(gasFeeEstimates.gasPrice);
+
+      log('Using gasPrice from estimate', maxFeePerGas);
+
+      return {
+        maxFeePerGas,
+      };
+    }
+  } catch (error) {
+    log('Failed to get estimate', error);
+  }
+
+  const gasPriceDecimal = (await query(new EthQuery(provider), 'gasPrice')) as
```

Comment:
It looks like the rest of this function will run if:

a) the above if-statements don't capture all cases of gasEstimateType OR
b) an error is caught.

Is that what is intended? 

I feel like we can make the control flow in this function more explicit. Ideas:
- Use a switch statement with a default case in the try block.
- Put the code that only runs if there is an error in the catch block.

Reply:
I had a bit of difficulty understanding the control flow that leads the 2nd case to the fallback logic.

From my perspective, it might be helpful to encapsulate the fallback logic into a separate function, say `getFallbackSuggestedGasFees()`. In each of the 3 cases that you list in your previous comment, as soon as we realize that we need to use the fallback logic, you can simply `return getFallbackSuggestedGasFees();` immediately.

This approach could make the code more readable and easier to maintain.